File: scenarios/v2v/Multi-ObjectTracking.py
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import math
from kalmanFilter import KalmanFilter
from itertools import permutations
import time

def generate_cost_matrix(pred_tracks, active_tracks, past_active_tracks, past_pred_tracks, detections, distance_weight=0.3, direction_weight=1.0):

    """
    Input: -pred_tracks: list of the all the positions of each track in the present, 
           -active_tracks: list of active tracks,
           -past_active_tracks: list of active tracks in the previous iteration,
           -past_pred_tracks: list of all the positions of each track in the previous iteration,
           -detections: list of detections,
           -distance_weight=0.3: weight that the distance will have to the cost,
           -direction_weight=0.3: weight that the direction will have to the cost
    Output: cost_matrix
    """

    n = len(active_tracks)
    m = len(detections)

    # inicializar cost_matrix com 0s
    cost_matrix = np.zeros((n, m))

    for i in range(n):        
        for j in range(m):
            pred_track_pos = pred_tracks[active_tracks[i]]  
            detection_pos = detections[j]
            past_pred_tracks_pos = past_pred_tracks[past_active_tracks[i]]
            
            # calcula a distancia entre as posiçoes atuais das tracks e das deteçoes 
            dx = pred_track_pos[0] - detection_pos[0]
            dy = pred_track_pos[1] - detection_pos[1]
            
            distance = np.sqrt(float(dx)**2 + float(dy)**2)
            ang_dif = 0
            if len(past_active_tracks)>0: # só é possivel calcular a direção se existir deteções passadas
                
                # calcula a direção entre as posiçoes passadas das tracks e das posiçoes presentes das tracks 
                delta_x = past_pred_tracks_pos[0] - pred_track_pos[0]
                delta_y = past_pred_tracks_pos[1] - pred_track_pos[1]

                angle1 = math.atan2(delta_y, delta_x)
                
                # calcula a direção entre as posiçoes autais das tracks e das deteçoes
                delta_x = pred_track_pos[0] - detection_pos[0]
                delta_y = pred_track_pos[1] - detection_pos[1]

                angle2 = math.atan2(delta_y, delta_x)
                
                # calcular a diferença absoluta entre os dois angulos
                ang_dif = np.abs(angle1-angle2)

            cost_matrix[i][j] = round(distance, 2)*distance_weight + round(ang_dif, 2)*direction_weight
    return cost_matrix

def optimal_assignment(cost_matrix):
    num_tracks = cost_matrix.shape[0]
    
    # Generate all possible assignments (permutations)
    indices = range(num_tracks)
    best_cost = float('inf')
    best_assignment = None

    for perm in permutations(indices):
        # Calculate the cost for this assignment
        total_cost = sum(cost_matrix[i, perm[i]] for i in indices)
        
        # Check if this is the best cost found
        if total_cost < best_cost:
            best_cost = total_cost
            best_assignment = perm

    return best_assignment


def main():
    # Load detection positions from the files (including confidence)
    # detections = pd.read_csv('car1_detection_pos3.txt')
    #positions = pd.read_csv('positions.txt')
    
    pred_tracks = {}   # map of predicted tracks
    past_pred_tracks = {}   # map of past predicted tracks
    tracks_count_id = 0 # counter of tracks used for id atribution
    active_tracks = []  # list with the ids of active tracks
    past_active_tracks = []   # list with the ids of active tracks in the past iteration
    kf_map = {}          # map with of Kalman filters associated to each track

    # Skip the first line (header)
    detections = detections.iloc[1:]
    #positions = positions.iloc[1:]

    # Sort detections by the timestamp
    # detections = detections.sort_values(by='timestamp')
    # Group detections by timestamp
    # grouped_detections = detections.groupby('timestamp')
    
    timestamp = time.time()

    for key in pred_tracks: #atualiza o array de tracks ativas, removendo as tracks que não sao atualizadas a mais de 5 segundos
        if timestamp - pred_tracks[key][2] > 5 and key in active_tracks:
            active_tracks.remove(key)
            if key in past_active_tracks:
                past_active_tracks.remove(key)
        
    pos = np.array([unique_group['x'], unique_group['y']]).T
        
    if len(active_tracks) == 0: # if we dont have any tracks yet, the detections will be the first element of the new tracks
            
        for i in range(len(pos)):
            pred_tracks[tracks_count_id] = (pos[i, 0], pos[i, 1], timestamp)
                
            active_tracks.append(tracks_count_id)
            kf_map[tracks_count_id] = KalmanFilter(0.05, 0, 0, 0, 0.3, 0.3)

            tracks_count_id += 1
                
    else:
        cost_matrix = generate_cost_matrix(pred_tracks, active_tracks, past_active_tracks, past_pred_tracks, pos)
        min_cost_indices = optimal_assignment(cost_matrix) # we calculate the otimal assignement
        """
        cost matrix:   
                        detections:
                      0    1    3
                0  [[0.8  0.3  0.9],          
        tracks: 1   [0.3  0.9  0.9],
                2   [0.9  0.8  0.3]]

        min_cost_indices: [1  0  2]
        
        track 0 -> detection 1
        track 1 -> detection 0
        track 2 -> detection 2
        """
            
        if len(active_tracks) == len(pos): #number of detections is equal to the number of active tracks

            if len(min_cost_indices) != len(set(min_cost_indices)): 
            # if min_cost_indices has repeated values, that means that a detection was atribuited to more than one track
                min_cost_indices = optimal_assignment(cost_matrix) # we calculate the otimal assignement
                
                #with the cost matrix, we can atribuit each detection to its track
            for i,value in enumerate(active_tracks):
                pred_tracks[value] = (pos[min_cost_indices[i], 0], pos[min_cost_indices[i], 1], timestamp)
                
                     
        elif len(active_tracks) < len(pos): #number of detections is greater than the number of active tracks
                
            # the existing tracks get updated with the new detection
            for i,value in enumerate(active_tracks):
                pred_tracks[value] = (pos[min_cost_indices[i], 0], pos[min_cost_indices[i], 1], timestamp)
 
            # a new track for each detection that doesnt have a track is added
            for i in range(len(pos) - len(active_tracks)):
                pred_tracks[tracks_count_id] = (pos[len(active_tracks)-1+i, 0], pos[len(active_tracks)-1+i, 1], timestamp)
                active_tracks.append(tracks_count_id)
                kf_map[tracks_count_id] = (KalmanFilter(0.05, 0, 0, 0, 0.3, 0.3))

                tracks_count_id += 1  
            
        else: # there are more tracks than detections, meaning that one track is over or hiden
            min_cost = 0
            tracks_aux = -1
            pos_aux = -1
        
            for j in range(len(pos)):
                min_cost = cost_matrix[0][j]
                pos_aux = j
                tracks_aux = 0
                for i in range(len(active_tracks)): # calculates which tracks have the lower cost for the detections
                    if cost_matrix[i][j] < min_cost:
                        min_cost = cost_matrix[i][j]
                        tracks_aux = i 
                            
                pred_tracks[active_tracks[tracks_aux]] = (pos[pos_aux, 0], pos[pos_aux, 1], timestamp)

    # Update each Kalman filter with the current position
    for at in active_tracks:

        prediction = kf_map[at].predict()  # Predict the next position with the KALMAN FILTER
        update = kf_map[at].update([pred_tracks[at][0], pred_tracks[at][1]])  # Update with the current position with the KALMAN FILTER
        
    past_pred_tracks = pred_tracks.copy()
    past_active_tracks = active_tracks.copy()

if __name__ == "__main__":
    main()

File: .Trash-1000/files/src/artery/application/MultiObjectTracking.cpp
#include <iostream>
#include <vector>
#include <cmath>
#include <map>
#include <set>
#include <algorithm>
#include <ctime>
#include <iterator>
#include <numeric>
#include <limits>
#include <tuple>
#include <array>
#include <functional>

#include "KalmanFilter.cc"


std::vector<std::vector<double>> generate_cost_matrix(
    const std::map<int, std::tuple<double, double, double>>& pred_tracks,
    const std::vector<int>& active_tracks,
    const std::vector<int>& past_active_tracks,
    const std::map<int, std::tuple<double, double, double>>& past_pred_tracks,
    const std::vector<std::array<double, 2>>& detections,
    double distance_weight = 0.3,
    double direction_weight = 1.0) {

    size_t n = active_tracks.size();
    size_t m = detections.size();

    std::vector<std::vector<double>> cost_matrix(n, std::vector<double>(m, 0.0));

    for (size_t i = 0; i < n; ++i) {
        for (size_t j = 0; j < m; ++j) {
            auto pred_track_pos = pred_tracks.at(active_tracks[i]);
            auto detection_pos = detections[j];
            auto past_pred_tracks_pos = past_pred_tracks.at(past_active_tracks[i]);

            double dx = std::get<0>(pred_track_pos) - detection_pos[0];
            double dy = std::get<1>(pred_track_pos) - detection_pos[1];

            double distance = std::sqrt(dx * dx + dy * dy);
            double ang_dif = 0.0;

            if (!past_active_tracks.empty()) {
                double delta_x = std::get<0>(past_pred_tracks_pos) - std::get<0>(pred_track_pos);
                double delta_y = std::get<1>(past_pred_tracks_pos) - std::get<1>(pred_track_pos);

                double angle1 = std::atan2(delta_y, delta_x);

                delta_x = std::get<0>(pred_track_pos) - detection_pos[0];
                delta_y = std::get<1>(pred_track_pos) - detection_pos[1];

                double angle2 = std::atan2(delta_y, delta_x);

                ang_dif = std::abs(angle1 - angle2);
            }

            cost_matrix[i][j] = std::round(distance * 100.0) / 100.0 * distance_weight + std::round(ang_dif * 100.0) / 100.0 * direction_weight;
        }
    }
    return cost_matrix;
}

std::vector<int> optimal_assignment(const std::vector<std::vector<double>>& cost_matrix) {
    size_t num_tracks = cost_matrix.size();
    std::vector<int> indices(num_tracks);
    std::iota(indices.begin(), indices.end(), 0);
    double best_cost = std::numeric_limits<double>::infinity();
    std::vector<int> best_assignment;

    do {
        double total_cost = 0.0;
        for (size_t i = 0; i < num_tracks; ++i) {
            total_cost += cost_matrix[i][indices[i]];
        }

        if (total_cost < best_cost) {
            best_cost = total_cost;
            best_assignment = indices;
        }
    } while (std::next_permutation(indices.begin(), indices.end()));

    return best_assignment;
}


int main() {
    std::map<int, std::tuple<double, double, double>> pred_tracks;
    std::map<int, std::tuple<double, double, double>> past_pred_tracks;
    std::map<int, int> pred_tracks_repeat;
    int tracks_count_id = 0;
    std::vector<int> active_tracks;
    std::vector<int> past_active_tracks;
    std::map<int, KalmanFilter> kf_map;

    std::vector<std::array<double, 2>> detections; // Load detection positions

    std::time_t timestamp = std::time(nullptr);

    for (auto key : pred_tracks) {
        if (timestamp - std::get<2>(key.second) > 5 && std::find(active_tracks.begin(), active_tracks.end(), key.first) != active_tracks.end()) {
            active_tracks.erase(std::remove(active_tracks.begin(), active_tracks.end(), key.first), active_tracks.end());
            if (std::find(past_active_tracks.begin(), past_active_tracks.end(), key.first) != past_active_tracks.end()) {
                past_active_tracks.erase(std::remove(past_active_tracks.begin(), past_active_tracks.end(), key.first), past_active_tracks.end());
            }
        }
    }

    std::vector<std::array<double, 2>> pos; // Populate with unique_group['x'], unique_group['y']

    if (active_tracks.empty()) {
        for (size_t i = 0; i < pos.size(); ++i) {
            pred_tracks[tracks_count_id] = std::make_tuple(pos[i][0], pos[i][1], timestamp);
            pred_tracks_repeat[tracks_count_id] = 1;

            active_tracks.push_back(tracks_count_id);
            kf_map[tracks_count_id] = KalmanFilter(0.05, 0, 0, 0, 0.3, 0.3);

            tracks_count_id++;
        }
    } else {
        auto cost_matrix = generate_cost_matrix(pred_tracks, active_tracks, past_active_tracks, past_pred_tracks, pos);
        auto min_cost_indices = optimal_assignment(cost_matrix);

        if (active_tracks.size() == pos.size()) {
            if (min_cost_indices.size() != std::set<int>(min_cost_indices.begin(), min_cost_indices.end()).size()) {
                min_cost_indices = optimal_assignment(cost_matrix);
            }

            for (size_t i = 0; i < active_tracks.size(); ++i) {
                pred_tracks[active_tracks[i]] = std::make_tuple(pos[min_cost_indices[i]][0], pos[min_cost_indices[i]][1], timestamp);

                if (pred_tracks_repeat.find(active_tracks[i]) != pred_tracks_repeat.end()) {
                    pred_tracks_repeat[active_tracks[i]]++;
                } else {
                    pred_tracks_repeat[active_tracks[i]] = 1;
                }
            }
        } else if (active_tracks.size() < pos.size()) {
            for (size_t i = 0; i < active_tracks.size(); ++i) {
                pred_tracks[active_tracks[i]] = std::make_tuple(pos[min_cost_indices[i]][0], pos[min_cost_indices[i]][1], timestamp);
                pred_tracks_repeat[active_tracks[i]]++;
            }

            for (size_t i = 0; i < pos.size() - active_tracks.size(); ++i) {
                pred_tracks[tracks_count_id] = std::make_tuple(pos[active_tracks.size() - 1 + i][0], pos[active_tracks.size() - 1 + i][1], timestamp);
                pred_tracks_repeat[tracks_count_id] = 1;
                active_tracks.push_back(tracks_count_id);
                kf_map[tracks_count_id] = KalmanFilter(0.05, 0, 0, 0, 0.3, 0.3);

                tracks_count_id++;
            }
        } else {
            double min_cost = 0;
            int tracks_aux = -1;
            int pos_aux = -1;

            for (size_t j = 0; j < pos.size(); ++j) {
                min_cost = cost_matrix[0][j];
                pos_aux = j;
                tracks_aux = 0;
                for (size_t i = 0; i < active_tracks.size(); ++i) {
                    if (cost_matrix[i][j] < min_cost) {
                        min_cost = cost_matrix[i][j];
                        tracks_aux = i;
                    }
                }

                pred_tracks[active_tracks[tracks_aux]] = std::make_tuple(pos[pos_aux][0], pos[pos_aux][1], timestamp);
                pred_tracks_repeat[active_tracks[tracks_aux]]++;
            }
        }
    }

    for (const auto& at : active_tracks) {
        auto prediction = kf_map[at].predict();
        kf_map[at].update({std::get<0>(pred_tracks[at]), std::get<1>(pred_tracks[at])});
    }

    past_pred_tracks = pred_tracks;
    past_active_tracks = active_tracks;
    return 0;
}